library/alloc/src/collections/vec_deque/spec_extend.rs

```diff
@@ -34,6 +35,66 @@ where
     }
 }
 
+impl<T, I, A: Allocator> SpecExtend<T, I> for VecDeque<T, A>
+where
+    I: TrustedLen<Item = T>,
+{
+    default fn spec_extend(&mut self, mut iter: I) {
+        // This is the case for a TrustedLen iterator.
+        let (low, high) = iter.size_hint();
+        if let Some(additional) = high {
+            debug_assert_eq!(
+                low,
+                additional,
+                "TrustedLen iterator's size hint is not exact: {:?}",
+                (low, high)
+            );
+            self.reserve(additional);
+
+            struct WrapAddOnDrop<'a, T, A: Allocator> {
+                vec: &'a mut VecDeque<T, A>,
+                written: usize,
+            }
+
+            impl<'a, T, A: Allocator> Drop for WrapAddOnDrop<'a, T, A> {
+                fn drop(&mut self) {
+                    self.vec.head = self.vec.wrap_add(self.vec.head, self.written);
+                }
+            }
+
+            let mut local_vec = WrapAddOnDrop { vec: self, written: 0 };
+
+            let head_room = local_vec.vec.cap() - local_vec.vec.head;
+            if additional <= head_room {
+                unsafe {
+                    local_vec.vec.write_iter(local_vec.vec.head, iter, &mut local_vec.written);
+                }
+            } else {
+                unsafe {
+                    local_vec.vec.write_iter(
+                        local_vec.vec.head,
+                        iter.by_ref().take(head_room),
```

Comment:
Ah, `by_ref` might interfere with getting any benefits from `for_each` because the implementation doesn't pass through most of the methods.

https://github.com/rust-lang/rust/blob/c08b235a5ce10167632bb0fddcd0c5d67f2d42e3/library/core/src/iter/traits/iterator.rs#L3776-L3792

You could try [ByRefSized](https://github.com/rust-lang/rust/blob/master/library/core/src/iter/adapters/by_ref_sized.rs#L7) instead, but you'll have to expose it as unstable feature from core to be available in std.

If that's too much work we can also skip that part, they improvement already looks pretty nice as it is.

Reply:
Both Compiler Explorer and the benchmarks agree it's worth it

https://rust.godbolt.org/z/vcsnM96oE

# Benchmarks

I don't think it will be worth to commit those two benchmarks because I had to modify the `Extend` implementation to always call the second branch. Code is:

```rust
#[bench]
fn bench_extend_chained_trustedlen(b: &mut Bencher) {
    let mut ring: VecDeque<u16> = VecDeque::with_capacity(1000);

    b.iter(|| {
        ring.clear();
        ring.extend(black_box((0..256).chain(768..1024)));
    });
}

#[bench]
fn bench_extend_chained_bytes(b: &mut Bencher) {
    let mut ring: VecDeque<u16> = VecDeque::with_capacity(1000);
    let input1: &[u16] = &[128; 256];
    let input2: &[u16] = &[255; 256];

    b.iter(|| {
        ring.clear();
        ring.extend(black_box(input1.iter().chain(input2.iter())));
    });
}
```

## SpecExtend `.for_each`

```
test vec_deque::bench_extend_chained_bytes      ... bench:       1,254 ns/iter (+/- 18)
test vec_deque::bench_extend_chained_trustedlen ... bench:         999 ns/iter (+/- 3)
```

## SpecExtend `.for_each` + `ByRefSized`

```
test vec_deque::bench_extend_chained_bytes      ... bench:         823 ns/iter (+/- 21)
test vec_deque::bench_extend_chained_trustedlen ... bench:          25 ns/iter (+/- 1)
```

Reply:
> I don't think it will be worth to commit those two benchmarks because I had to modify the Extend implementation to always call the second branch. 

Have you tried lifting the first part out of the branch so it only does the 2nd part conditionally?

Reply:
Done